Segfault investigation

hsorby opened the pull request:
Fixing segfaulting issue with some models when using the libCellML library in a stand-alone application.

Addresses #432. 

This pull request is built on top of other pending pull requests and should be considered after they have been merged.

hsorby committed:
Only look at cellml namespace attributes on cn nodes, ignore others and let the MathML DTD pick them up.

hsorby committed:
Investigating segfault arising from model.

hsorby committed:
Merge branch 'export_compiler_features_i428' into maths_validation_i424

hsorby committed:
Add a little bit more to dbg class.  Move to using specialisations to avoid decaying array references.

hsorby committed:
Add an isolated test of tutorial 4.

hsorby committed:
Add segfaulting math encapsulation.

hsorby committed:
Merge branch 'segfault_investigation' of github.com:hsorby/libcellml into segfault_investigation

hsorby committed:
Merge branch 'export_compiler_features_i428' into segfault_investigation

hsorby committed:
Further investigations into the segfault.

hsorby committed:
Merge branch 'develop' into segfault_investigation

hsorby committed:
Remove debugging code. When looking for node first child only consider nodes that are not empty.

hsorby committed:
Merge branch 'segfault_investigation' of github.com:hsorby/libcellml into segfault_investigation

hsorby committed:
Format code with clang format.

hsorby committed:
Fix firstChild code to please clang-tidy.

hsorby committed:
Restrict test to what was causing the segfault and reference issue which found the problem.

agarny commented on src/generator.cpp:
```diff
@@ -553,8 +553,8 @@ struct Generator::GeneratorImpl
-    std::list<GeneratorInternalVariablePtr> mInternalVariables;
-    std::list<GeneratorEquationPtr> mEquations;
+    std::list<GeneratorInternalVariablePtr> mInternalVariables = {};
+    std::list<GeneratorEquationPtr> mEquations = {};
```
This shouldn't be needed...?

hsorby replied:
No it's not I thought clang tidy might remind me to remove it.

hsorby committed:
Merge in changes from maths_validation_i424 fixing merge conflicts.

hsorby committed:
Remove cmake/test_libxml2.cpp file.

hsorby committed:
Remove unnecessary vector initialisation.

hsorby committed:
Merge branch 'develop' into segfault_investigation

hsorby committed:
Change Generator to make use of owningModel function insteam of dynamic casting parent pointer.

hsorby committed:
Merge in changes from unfound_units_i430 fixing merge conflicts.

hsorby committed:
Merge in changes from prime/develop fixing conflicts.

hsorby committed:
Apply clang format fixes.

hsorby committed:
Merge branch 'develop' into segfault_investigation

hsorby committed:
Merge branch 'develop' into segfault_investigation

hsorby committed:
Merge in changes from multiplier_report_equiv_units_i381 fixing conflicts.

hsorby committed:
Merge branch 'segfault_investigation' of github.com:hsorby/libcellml into segfault_investigation

hsorby committed:
Prepend errors from LibXml2 with 'LibXml2 error: '.

hsorby committed:
Merge branch 'develop' into segfault_investigation

hsorby committed:
Merge in changes from prime/develop fixing conflicts.

hsorby committed:
Remove ignore utilities.h when using llvm coverage.

hsorby commented:
This pull request is now ready for review.

kerimoyle commented:
Looks good to me :) The test from #431 is also included and passing, so this PR will close that issue too.

kerimoyle commented on tests/isolated/generator.cpp:
```diff
@@ -0,0 +1,211 @@
+/*
```
I don't understand how this is more "isolated" than the other tests?  It's still linking to the library in the same way, so it doesn't really test anything different?

kerimoyle commented on tests/parser/parser.cpp:
```diff
@@ -37,11 +37,11 @@ TEST(Parser, invalidXMLElements)
-        "Opening and ending tag mismatch: Wizard line 5 and SomeGuyWithAStaff.",
-        "Opening and ending tag mismatch: Elf line 6 and fellows.",
-        "Premature end of data in tag fellowship line 2.",
+        "LibXml2 error: Opening and ending tag mismatch: Dwarf line 3 and ShortGuy.",
```
I like the starting tags!  Makes it a bit more contextual when you're trying to interpret the error :)

agarny committed:
Generator: put back some empty lines.

If we are to remove empty lines, which to me makes the code easier to read, then we should do it throughout the file, not only at some "random" places.

agarny committed:
XmlNode: simplified XmlNode::firstChild().

agarny committed:
Isolated generator test: some minor cleaning up of the code.

To make it more consistent with the rest of our codebase.

agarny committed:
Validator test: cleaned up and renamed the mathInEncapsulatedComponentSegfault test.
